backends/p4tools/common/compiler/convert_varbits.h

```diff
@@ -0,0 +1,34 @@
+#ifndef COMMON_COMPILER_CONVERT_VARBITS_H_
+#define COMMON_COMPILER_CONVERT_VARBITS_H_
+
+#include "frontends/common/resolveReferences/referenceMap.h"
+#include "frontends/p4/typeMap.h"
+#include "ir/ir.h"
+#include "lib/null.h"
+
+namespace P4Tools {
+
+/// Converts all existing Type_Varbit types in the program into a custom Size_Type_Varbit type.
```

Comment:
Nothing precludes the runtime implementation of varbits from doing this. In fact, it does in bmv2.

Reply:
We need a varbit type that retains the original max-width of the varbit and also contains information about the current assigned width. I do not think this is possible with the current varbit type? How does BMv2 do it? It would simplify the code a lot.

Reply:
Quoting from https://p4.org/wp-content/uploads/2022/07/P4-16-spec.html#sec-varbit-string

> Each occurrence of the type varbit has a statically-declared maximum width, as well as a dynamic width, which must not exceed the static bound. Prior to initialization a variable-size bit-string has an unknown dynamic width.
> 
> Variable-length bit-strings support a limited set of operations:
> 
>     Assignment to another variable-sized bit-string. The target of the assignment must have the same static width as the source. When executed, the assignment sets the dynamic width of the target to the dynamic width of the source.
>     Comparison for equality or inequality with another varbit field. Two varbit fields can be compared only if they have the same type. Two varbits are equal if they have the same dynamic width and all the bits up to the dynamic width are the same. 

Isn't this exactly what you need?

Reply:
Yes, afaik the `IR::Type_Varbits` does only have one size field. Even if we generated a new Type_Varbits, we would lose information about the original size and the types are not equivalent.

Reply:
The max size is compile-time information.
The compiler is not concerned with the dynamic size, that only exists at runtime.
I still fail to see why you need your encoding.

Reply:
Our interpreter assigns a dynamic size [here](https://github.com/p4lang/p4c/pull/3495/files#diff-ed409c5ff044fe5d8cafe027e7ef5007f433def7e39bd28974d9f3253c7be1aeR36) when we call an extract. At this point the varbit will have both information about the original maximum size and the dynamic size. We use the maximum size as means of comparison to avoid type mismatch issues.  
https://github.com/p4lang/p4c/pull/3495/files#diff-bfbb8698949a4875ee7cd51ba9a17c7a58f9a7a6a1f4542455a3b551f483ba30R18
(I will need to update that comment)